src/bare-metal/microcontrollers/examples/src/bin/hal.rs

```diff
@@ -0,0 +1,44 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// ANCHOR: Example
+#![no_main]
+#![no_std]
+
+extern crate panic_halt as _;
+
+use cortex_m_rt::entry;
+use nrf52833_hal::{
+    gpio::{p0, Level},
+    pac::Peripherals,
+    prelude::*,
+};
+
+#[entry]
+fn main() -> ! {
+    let p = Peripherals::take().unwrap();
+
+    // GPIO port 0.
+    let gpio0 = p0::Parts::new(p.P0);
+
+    // Configure GPIO0 pins 21 and 28 as push-pull outputs.
+    let mut col1 = gpio0.p0_28.into_push_pull_output(Level::High);
+    let mut row1 = gpio0.p0_21.into_push_pull_output(Level::Low);
+
+    // Set pin 28 high and pin 21 low to turn the LED on.
+    col1.set_low().unwrap();
+    row1.set_high().unwrap();
+
+    loop {}
```

Comment:
What is the behavior if we don't loop here? Perhaps a slide on the lifecycle of a program running on a microcontroller would be good?

Reply:
The `cortex_m_rt::entry` macro requires that the function have type `fn() -> !`, because returning to the reset handler doesn't make sense.